alf/networks/preprocessor_networks.py

```diff
@@ -0,0 +1,137 @@
+                arg is helpful if you want to combine inputs by configuring a
+                gin file without changing the code.
+            name (str): name of the network
+        """
+        super().__init__(input_tensor_spec, name)
+
+        # make sure the network holds the parameters of any trainable input
+        # preprocessor
+        self._input_preprocessor_modules = nn.ModuleList()
+
+        def _get_preprocessed_spec(preproc, spec):
+            if not isinstance(preproc, InputPreprocessor):
+                # In this case we just assume the spec won't change after the
+                # preprocessing. If it does change, then you should consider
+                # defining an `InputPreprocessor` instead.
+                return spec
+            return preproc(spec)
+
+        self._input_preprocessors = None
+        if input_preprocessors is not None:
+            input_tensor_spec = alf.nest.map_structure(
+                _get_preprocessed_spec, input_preprocessors, input_tensor_spec)
+
+            def _return_or_copy_preprocessor(preproc):
+                if preproc is None:
+                    # allow None as a placeholder in the nest
+                    preproc = math_ops.identity
+                elif isinstance(preproc, InputPreprocessor):
+                    preproc = preproc.copy()
+                    self._input_preprocessor_modules.append(preproc)
+                return preproc
+
+            self._input_preprocessors = alf.nest.map_structure(
+                _return_or_copy_preprocessor, input_preprocessors)
+
+        self._preprocessing_combiner = preprocessing_combiner
+        if alf.nest.is_nested(input_tensor_spec):
+            assert preprocessing_combiner is not None, \
+                ("When a nested input tensor spec is provided, a input " +
+                "preprocessing combiner must also be provided!")
+            input_tensor_spec = preprocessing_combiner(input_tensor_spec)
+        else:
+            assert isinstance(input_tensor_spec, TensorSpec), \
+                "The spec must be an instance of TensorSpec!"
+            self._preprocessing_combiner = math_ops.identity
+
+        # This input spec is the final resulting spec after input preprocessors
+        # and the nest combiner.
+        self._processed_input_tensor_spec = input_tensor_spec
+
+    def forward(self, inputs, state=(), min_outer_rank=1, max_outer_rank=1):
+        """Preprocessing nested inputs.
+
+        Args:
+            inputs (nested Tensor): inputs to the network
+            state (nested Tensor): RNN state of the network
+            min_outer_rank (int): the minimal outer rank allowed
+            max_outer_rank (int): the maximal outer rank allowed
+        Returns:
+            Tensor: tensor after preprocessing.
+        """
+        if self._input_preprocessors:
+            inputs = alf.nest.map_structure(
+                lambda preproc, tensor: preproc(tensor),
```

Comment:
if preproc is Network, we should use preproc(tensor)[0]

Reply:
If it's always stateless, then combiner will ignore the empty state.

Reply:
I see.

Reply:
Will improve on this.